### ISSUE TITLE
Create new npm release

### ISSUE THREAD

**Issue by tony-scio** [reduxjs/redux-thunk#299]
# Feature Request

There are a lot of good updates since the last npm release over 2 years ago. https://github.com/reduxjs/redux-thunk/compare/v2.3.0...master

It'd be really great if whoever has access could cut a new release.

**Comment by Ryunaq** [comment]
Second this, mainly the changes to the types definitions would have helped me a lot.

**Comment by eaglus** [comment]
@timdorr  @gaearon @aikoven  Please publish a latest version of this package. We use our private npm package based on latest master of redux-think, and dream to have a latest official version, and drop out private package

**Comment by gaearon** [comment]
@tony-scio Do you want to cut npm releases? I'm happy to add as an owner.

**Comment by markerikson** [comment]
FWIW, the last time Tim and I chatted about this, he was concerned that the current types may not be sufficiently correct or something along those lines.  Tim, can you clarify what those concerns are atm?

**Comment by timdorr** [comment]
Yes, I'm primarily concerned that we're using a number of overloads to apply the thunk types, rather than preferring the middleware approach. This means that the Dispatch and bindActionCreators types get overridden if the module is included, even if it's not used. 

This also creates a problem for using multiple middlewares on the same store, where they may get clobbered by this overload. I'm not sure of the priority order on overloads vs. directly applied types, but that uncertainty makes me wary. 

For reference, here are the PRs with type changes since 2.3.0:

#216
#219 
#224
#243 
#255
#260
#263
#278

I have some regret about merging in #224 and #278, since they add the overloads. They also tie to the specific types we have in Redux 4.x, and will break under 5.0.

**Comment by tony-scio** [comment]
@gaearon I'm probably not yet well qualified to be responsible for redux-thunk releases as I haven't developed in this library. I'm just a consumer of the library who is still very much looking forward to the next release.

**Comment by flymoondust** [comment]
I'm also looking forward the next release。My company has standard for picking third part library,Now for 2 years no new release, I may be forced to pick alternative library, but I don't know any other library.

**Comment by timdorr** [comment]
Just keep in mind the only thing that has changed in that time is the build system and the TypeScript types. The code of the actual middleware hasn't changed in 4 years, and that was only to add extraArgument.

**Comment by tony-scio** [comment]
@timdorr apologies for the naive question, but why not make a release branch and, on it, back out the 2 commits you mentioned that you regret? One of the other type improvements would have saved me a prod bug. Seems unfortunate to hold them back from users.

**Comment by gregtatum** [comment]
Just chiming in to say that I've put lots of hours over the years into trying to type thunks in both Flow types and TypeScript, so having some kind of type-safe solution that is type-correct for connected components, even if not ergonomic would be a huge windfall. I'm more concerned about catching bugs, rather than having an ergonomic API. This typing doesn't appear to be a solved problem with any kind of official recommendation, even if it's a "this is not great, but it works".

That said, thanks for the maintenance work, I find redux-thunk essential for the apps I maintain.

**Comment by GLObus303** [comment]
Hey, any plans on releasing the fixes?

**Comment by flymoondust** [comment]
Hey, any plans on releasing the fixes?

**Comment by Philipp91** [comment]
> Yes, I'm primarily concerned that we're using a number of overloads

Fair. https://github.com/microsoft/TypeScript/issues/14107 is annoying, I contributed a [workaround](https://github.com/reduxjs/redux-thunk/pull/255) here, but the root problem (the overloads) remains. [Elsewhere](https://github.com/reduxjs/redux-thunk/pull/255) they recommend:

> you should never write a series of overloads which have an equivalent representation in union types.

Still, as it's been three years, I'm in favor of a release or some move in _some_ direction. I currently use the latest release `"redux-thunk": "^2.3.0",`, and I'm getting plenty of TypeScript errors due to #248 unless I manually patch [the workaround](https://github.com/reduxjs/redux-thunk/issues/288#issuecomment-744076132).

**Comment by markerikson** [comment]
As mentioned in that other thread, using our official Redux Toolkit package (which integrates thunks into `configureStore`) should result in correct types behavior, and you should be using RTK anyway:

https://redux-toolkit.js.org/api/configureStore

**Comment by Philipp91** [comment]
Here's how to reproduce the problem with Redux Toolkit (and applying the aforementioned workaround fixes it here too):

```
// type-tests/files/createMixedThunk.typetest.ts

import {AnyAction, PayloadAction} from '@reduxjs/toolkit'
import {ThunkAction, ThunkDispatch} from 'redux-thunk'

const defaultDispatch = (() => {
}) as ThunkDispatch<{}, any, AnyAction>
type AppAction = PayloadAction<number> | ThunkAction<void, {}, any, AnyAction>;

;(async function () {
    const createAction = (id: number): PayloadAction<number> => ({type: '', payload: id});
    const createAsync = (id: number): ThunkAction<void, {}, any, AnyAction> => async (dispatch) => {
        dispatch(createAction(id));
    };
    const createMixed = (choose: boolean, id: number) => choose ? createAction(id) : createAsync(id);

    defaultDispatch(createAction(42)); // Works fine
    defaultDispatch(createAsync(42)); // Works fine
    const mixed: AppAction = createMixed(true, 42);
    defaultDispatch(mixed);
    //             ^^^^^^^ -- Argument types do not match parameters
})();
```

Basically the `mixed: AppAction` variable is either a plain old action, or it's a thunk. The reason we don't know and have to use a type union is because of some application logic (not part of the example code above) that passes the action around (which is why the `AppAction` type exists to begin with, though the error message is the same without that type). Imagine some helper functions/classes with signatures that accept or produce `AppAction` instances, so there's really no way to know which of the two action types it is.
